src/D2L.CodeStyle.Analyzers/Async/Generator/SyncGenerator.cs

```diff
@@ -0,0 +1,182 @@
+﻿using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace D2L.CodeStyle.Analyzers.Async.Generator;
+
+[Generator]
+internal sealed partial class SyncGenerator : IIncrementalGenerator {
+	public void Initialize( IncrementalGeneratorInitializationContext context ) {
+		// Collect all the methods we want to generate sync versions for
+		// individually (for better incremental builds)
+		IncrementalValuesProvider<(MethodDeclarationSyntax, Compilation)> methodsToGenerate =
+			context.SyntaxProvider
+			   .CreateSyntaxProvider( IsInterestingLookingMethod, ExtractSyntax )
+			   .Where( static m => m != null )
+			   .Combine( context.CompilationProvider )
+			   .WithComparerThatIgnoresCompilation()!;
+
+		// Do the generation per method
+		IncrementalValuesProvider<MethodGenerationResult> generatedMethods =
+			methodsToGenerate
+				.Select( GenerateSyncMethod )
+				// Filter out things we simply couldn't generate anything for.
+				.Where( static mr => mr.HasValue )
+				.Select( static ( mr, _ ) => mr!.Value );
+
+		// Collect and generate the output files
+		//
+		// Opportunity for better caching here -- changing one source file
+		// should only impact one output file. Currently we only generate the
+		// methods when their source file changed, but we re-collect all of
+		// that when any file with [GenerateSync] has changed.
+		IncrementalValuesProvider<FileGenerationResult> generatedFiles =
+			generatedMethods.Collect()
+				.SelectMany( GenerateFiles );
+
+		context.RegisterSourceOutput( generatedFiles, WriteFiles );
+	}
+
+	private static bool IsInterestingLookingMethod(
+		SyntaxNode node,
+		CancellationToken _
+	) {
+		if( node is not MethodDeclarationSyntax method ) {
+			return false;
+		}
+
+		// We only care about async methods
+		if( !method.Modifiers.Any( t => t.IsKind( SyntaxKind.AsyncKeyword ) ) ) {
+			return false;
+		}
+
+		// The ones we care about have to have a particular attribute -- so
+		// perk up when we see _any_ attributes (we don't have a SemanticModel
+		// in this method.)
+		return method.AttributeLists.Any();

```

Comment:
Could consider doing a quick check for `attribute.Name.ToString()` being equal to one of `"GenerateSync"` or `"GenerateSyncAttribute"` before going to the model to verify in the transform. Believe that's what Jeff has done the the DtoGenerator.

It's more restrictive than only-model (requiring specific strings in the syntax), but at least still correct by backing up the string check with the symbol lookup later.

Reply:
I mostly based this off the Microsoft ones. The predicate is just supposed to be a cheap filter.
Doing it with strings is less correct because of aliases (`using X = D2L.CodeStyle.Annotations.GenerateSync`).
I don't care that much but I figure that's why it's typically done this way.

Reply:
Actually no, Jeff just does `unqualifiedNames.Contains( unqualifiedNameOfAttribute )`, so not less restrictive. That could be an EndsWith. https://github.com/Brightspace/D2L.CodeStyle/blob/master/src/D2L.CodeStyle.Analyzers/Language/AttributeAliasesAnalyzer.cs being added makes that ultimately good and correct.

Reply:
Example predicate: https://github.com/dotnet/runtime/blob/80ba698a41cafa092a12a18119ae4a1c57347d64/src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Parser.cs#L24-L26

Reply:
It's only correct with that analyzer, but I don't know that I'd call it _good_.
Anyway, I'm just doing what others do here.

Reply:
Would mostly like to see _us_ do it one way. I'm happy to defer that to follow-ups in either direction though.